packages/dids/src/did-uri.ts

```diff
@@ -0,0 +1,186 @@
+   *
+   * @example
+   * h4d3ixkwt6q5a455tucw7j14jmqyghdtbr6cpiz6on5oxj5bpr3o
+   */
+  id: string;
+
+  /**
+      * Optional path component of the DID URI.
+      *
+      * @example
+      * did:web:tbd.website/path
+      */
+  path?: string;
+
+  /**
+      * Optional query component of the DID URI.
+      *
+      * @example
+      * did:web:tbd.website?versionId=1
+      */
+  query?: string;
+
+  /**
+   * Optional fragment component of the DID URI.
+   *
+   * @example
+   * did:web:tbd.website#key-1
+   */
+  fragment?: string;
+
+  /**
+    * Optional query parameters in the DID URI.
+    *
+    * @example
+    * did:web:tbd.website?service=files&relativeRef=/whitepaper.pdf
+    */
+  params?: Record<string, string>;
+
+  /**
+   * Constructs a new `DidUri` instance.
+   *
+   * @param params - An object containing the parameters to be included in the DID URI.
+   * @param params.method - The name of the DID method.
+   * @param params.id - The DID method identifier.
+   * @param params.path - Optional. The path component of the DID URI.
+   * @param params.query - Optional. The query component of the DID URI.
+   * @param params.fragment - Optional. The fragment component of the DID URI.
+   * @param params.params - Optional. The query parameters in the DID URI.
+   */
+  constructor({ method, id, path, query, fragment, params }: {
+    method: string,
+    id: string,
+    path?: string,
+    query?: string,
+    fragment?: string,
+    params?: Record<string, string>
+  }) {
+    this.uri = `did:${method}:${id}`;
+    this.method = method;
+    this.id = id;
+    this.path = path;
+    this.query = query;
+    this.fragment = fragment;
+    this.params = params;
+  }
+
+  /**
+   * Parses a DID URI string into its individual components.
+   *
+   * @example
+   * ```ts
+   * const didUri = DidUri.parse('did:example:123?service=agent&relativeRef=/credentials#degree');
+   *
+   * console.log(didUri.uri)      // Output: 'did:example:123'
+   * console.log(didUri.method)   // Output: 'example'
+   * console.log(didUri.id)       // Output: '123'
+   * console.log(didUri.query)    // Output: 'service=agent&relativeRef=/credentials'
+   * console.log(didUri.fragment) // Output: 'degree'
+   * console.log(didUri.params)   // Output: { service: 'agent', relativeRef: '/credentials' }
+   * ```
+   *
+   * @params didUriString - The DID URI string to be parsed.
+   * @returns A `DidUri` object representing the parsed DID URI, or `null` if the input string is not a valid DID URI.
+   */
+  static parse(didUriString: string): DidUri | null {
+    // Return null if the input string is empty or not provided.
+    if (!didUriString) return null;
+
+    // Execute the regex pattern on the input string to extract URI components.
+    const match = DidUri.DID_URI_PATTERN.exec(didUriString);
+
+    // If the pattern does not match, or if the required groups are not found, return null.
+    if (!match || !match.groups) return null;
```

Comment:
No error message or exception? Just returning null if did does not parse?

Reply:
Yes, this is intentional.  It matches the behavior of the old `parseDid` function it replaces and is checked appropriately for null everywhere it is used.  There are typically specific errors that are returned in DID Resolution Results and this simplifies the logic to return errors without having to add countless try/catch.